flink-end-to-end-tests/flink-end-to-end-tests-common-elasticsearch/src/main/java/org/apache/flink/streaming/tests/ElasticsearchTestEmitter.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.flink.streaming.tests;
+
+import org.apache.flink.api.connector.sink2.SinkWriter;
+import org.apache.flink.connector.elasticsearch.sink.ElasticsearchEmitter;
+import org.apache.flink.connector.elasticsearch.sink.RequestIndexer;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** Test emitter for performing ElasticSearch indexing requests. */
+public interface ElasticsearchTestEmitter extends ElasticsearchEmitter<KeyValue<Integer, String>> {
+
+    @Override
+    default void emit(
+            KeyValue<Integer, String> element, SinkWriter.Context context, RequestIndexer indexer) {
+        addUpsertRequest(indexer, element);
```

Comment:
I think it would be good to revisit the current `TestEmitter` structure. You can replace the inheritance with composition by only implementing one class that has only a constructor parameter to build the `UpdateRequest`. Currently, `emit` is marked as default, although all implementors override it, and sharing createDoc via a default method also feels strange because it could be a static method.

Reply:
Fixed `emit` and `createDoc`, but I do not quite get which composition structure you propose.

Reply:
Maybe I missed your change but `emit` looks similar to before. My idea is basically to only have one `ElasticsearchTestEmitter` implementing the emitter interface, and this class is used by elasticsearch6 and elasticsearch7. Both modules pass the needed information to construct the request in the constructor of the `ElasticsearchTestEmitter`. There also are still conflicts with the master branch.

Reply:
Thanks for the clarification. Please let me know if the changes from the topmost commit are aligned with your idea.